### PR TITLE
Made grammar object optional

### LINKED ISSUE #299: Make `grammar` object optional
Make the `grammar` top-level object optional to enable libraries of grammar rules that do not specify a grammar, but can be reused elsewhere.

### PR DESCRIPTION
**Pull request by jbicker**
Grammar top level object is optional now.
However if it is missing in grammar which is configured in langium-config.json then langium generator fails.

Fixes #299

### CODE REVIEW

**Review by msujew — COMMENTED** [review]
The changes look good at first glance. I'll take a closer look at it later.

On another note, do you mind running the langium generator on the arithmetics and statemachine example as well? Their `grammar.ts` should be updated before merging this PR.

File: packages/langium-cli/src/generate.ts [hunk 1]
```diff
@@ -141,12 +141,17 @@ export async function generate(config: LangiumConfig, options: GenerateOptions):
         const buildResult = all.get(absGrammarPath);
         if (buildResult) {
             const grammar = buildResult.document.parseResult.value as Grammar;
+            if(!grammar.isDeclared) {
+                log('error', options, `The grammar top level object is missing in ${absGrammarPath}`.red);
```
Comment (spoenemann) [inline packages/langium-cli/src/generate.ts:145]: Proposal: change the message to "${absGrammarPath}: The entry grammar must start with the 'grammar' keyword."

**Review by spoenemann — COMMENTED** [review]
Please add a validation check to `langium-grammar-validator.ts`: A rule marked as `entry` is illegal if the containing grammar is not declared.

File: packages/langium/src/grammar/langium-grammar-validator.ts [hunk 2]
```diff
@@ -97,13 +97,15 @@ export class LangiumGrammarValidator {
...
-        if (entryRules.length === 0) {
+        if (grammar.isDeclared && entryRules.length === 0) {
             const possibleEntryRule = grammar.rules.find(e => ast.isParserRule(e) && !isDataTypeRule(e));
             if (possibleEntryRule) {
                 accept('error', 'The grammar is missing an entry parser rule. This rule can be an entry one.', { node: possibleEntryRule, property: 'name', code: IssueCodes.EntryRuleTokenSyntax });
             } else {
                 accept('error', 'This grammar is missing an entry parser rule.', { node: grammar, property: 'name' });
             }
+        } else if(!grammar.isDeclared && entryRules.length === 1) {
+            accept('error', 'A rule marked as entry is illegal if the containing grammar is not declared.', { node: entryRules[0], property: 'name' });
```
Comment (msujew) [inline packages/langium/src/grammar/langium-grammar-validator.ts:108]: Minor: This should be performed for every entry rule and run in addition to the other entry rule validation (currently is it running exclusively).

Additionally, minor wording improvement:

```suggestion
            accept('error', 'Cannot declare entry rules for unnamed grammars.', { node: entryRules[0], property: 'name' });
```

**Review by msujew — APPROVED** [review]

**Pull request merged**